File: csf_tz/csf_tz/report/depreciation_transaction_summary/depreciation_transaction_summary.py
# Copyright (c) 2025, Aakvatech and contributors
# For license information, please see license.txt
import frappe
from frappe import _
from frappe.query_builder import DocType
from frappe.query_builder.functions import Sum


def execute(filters=None):
    if not filters:
        return [], []
    
    columns = get_columns()
    data = get_data(filters)
    
    return columns, data


def get_columns():
    return [
        {
            "fieldname": "asset",
            "label": _("Asset"),
            "fieldtype": "Link",
            "options": "Asset",
            "width": 120
        },
        {
            "fieldname": "asset_name",
            "label": _("Asset Name"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "asset_category",
            "label": _("Asset Category"),
            "fieldtype": "Link",
            "options": "Asset Category",
            "width": 120
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 100
        },
        {
            "fieldname": "purchase_date",
            "label": _("Purchase Date"),
            "fieldtype": "Date",
            "width": 100
        },
        {
            "fieldname": "opening_wdv",
            "label": _("Opening WDV"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "depreciation_amount",
            "label": _("Depreciation Amount"),
            "fieldtype": "Currency",
            "width": 140
        },
        {
            "fieldname": "depreciation_date",
            "label": _("Depreciation Date"),
            "fieldtype": "Date",
            "width": 140
        },
        {
            "fieldname": "depreciation_entry",
            "label": _("Depreciation Entry"),
            "fieldtype": "Link",
            "options": "Journal Entry",
            "width": 140
        },
        {
            "fieldname": "total_accumulated_depreciation",
            "label": _("Current Period Depreciation"),
            "fieldtype": "Currency",
            "width": 180
        },
        {
            "fieldname": "closing_wdv",
            "label": _("Closing WDV"),
            "fieldtype": "Currency",
            "width": 120
        },

        {
            "fieldname": "rate_of_depreciation",
            "label": _("Rate of Depreciation"),
            "fieldtype": "Percent",
            "width": 150
        }
    ]


def get_data(filters):
    data = []

    # Fetch depreciation accounts using QB
    Account = DocType("Account")
    depreciation_accounts = (
        frappe.qb.from_(Account)
        .select(Account.name)
        .where(Account.account_type == "Depreciation")
        .run(pluck=True)
    )

    if not depreciation_accounts:
        return data

    # Base filters for GL Entry
    filters_data = {
        "company": filters.get("company"),
        "posting_date": ["between", [filters.get("from_date"), filters.get("to_date")]],
        "against_voucher_type": "Asset",
        "account": ["in", depreciation_accounts],
        "is_cancelled": 0,
    }

    # Add asset filter if provided
    if filters.get("asset"):
        filters_data["against_voucher"] = filters.get("asset")

    # Fetch assets based on asset category using QB
    if filters.get("asset_category"):
        Asset = DocType("Asset")
        assets = (
            frappe.qb.from_(Asset)
            .select(Asset.name)
            .where(
                (Asset.asset_category == filters.get("asset_category"))
                & (Asset.docstatus == 1)
            )
            .run(pluck=True)
        )
        filters_data["against_voucher"] = ["in", assets]

    # Fetch GL Entries using QB
    GLEntry = DocType("GL Entry")
    gl_entries = (
        frappe.qb.from_(GLEntry)
        .select(
            GLEntry.against_voucher,
            GLEntry.debit_in_account_currency.as_("depreciation_amount"),
            GLEntry.voucher_no,
            GLEntry.posting_date,
        )
        .where(
            (GLEntry.company == filters.get("company"))
            & (GLEntry.posting_date[filters.get("from_date") : filters.get("to_date")])
            & (GLEntry.against_voucher_type == "Asset")
            & (GLEntry.account.isin(depreciation_accounts))
            & (GLEntry.is_cancelled == 0)
        )
        .orderby(GLEntry.against_voucher, GLEntry.posting_date)
        .run(as_dict=True)
    )

    if not gl_entries:
        return data

    # Get unique assets from GL Entries
    assets = list(set(d["against_voucher"] for d in gl_entries))

    # Fetch asset details using QB
    assets_details = get_assets_details(assets, filters)

    # Process GL Entries
    for d in gl_entries:
        asset_data = assets_details.get(d["against_voucher"])
        if asset_data:
            if not asset_data.get("accumulated_depreciation_amount"):
                asset_data["accumulated_depreciation_amount"] = 0

            # Calculate opening WDV
            asset_data["opening_wdv"] = (
                asset_data["gross_purchase_amount"]
                - asset_data["accumulated_depreciation_amount"]
            )

            # Increment accumulated depreciation
            asset_data["accumulated_depreciation_amount"] += d["depreciation_amount"]

            # Rename the key after incrementing
            asset_data["total_accumulated_depreciation"] = asset_data.pop("accumulated_depreciation_amount")

            # Calculate closing WDV
            asset_data["closing_wdv"] = (
                asset_data["gross_purchase_amount"]
                - asset_data["total_accumulated_depreciation"]
            )

            # Prepare the row for the report
            row = asset_data.copy()
            row.update(
                {
                    "depreciation_amount": d["depreciation_amount"],
                    "depreciation_date": d["posting_date"],
                    "depreciation_entry": d["voucher_no"],
                }
            )

            # Append the row to the data list
            data.append(row)

    return data


def get_assets_details(assets, filters):
    assets_details = {}

    # Fetch asset details using QB
    Asset = DocType("Asset")
    AssetFinanceBook = DocType("Asset Finance Book")

    assets_data = (
        frappe.qb.from_(Asset)
        .left_join(AssetFinanceBook)
        .on(Asset.name == AssetFinanceBook.parent)
        .select(
            Asset.name.as_("asset"),
            Asset.asset_name,
            Asset.gross_purchase_amount,
            Asset.asset_category,
            Asset.status,
            Asset.purchase_date,
            Asset.cost_center,
            AssetFinanceBook.rate_of_depreciation,
        )
        .where(Asset.name.isin(assets))
        .run(as_dict=True)
    )

    for asset in assets_data:
        assets_details[asset["asset"]] = asset
        
        if not asset.get("rate_of_depreciation"):
            asset["rate_of_depreciation"] = 0

    return assets_details

File: csf_tz/csf_tz/report/depreciation_transaction_summary/depreciation_transaction_summary.js
// Copyright (c) 2025, Aakvatech and contributors
// For license information, please see license.txt
frappe.query_reports["Depreciation Transaction Summary"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "periodicity",
			label: __("Periodicity"),
			fieldtype: "Select",
			options: [
				{ value: "Monthly", label: __("Monthly") },
				{ value: "Quarterly", label: __("Quarterly") },
				{ value: "Yearly", label: __("Yearly") }
			],
			default: "Monthly",
			reqd: 1,
			on_change: function() {
				let periodicity = frappe.query_report.get_filter_value('periodicity');
				let today = frappe.datetime.get_today();
				
				if(periodicity === "Quarterly") {
					// Get current quarter
					let quarter = Math.floor((frappe.datetime.str_to_obj(today).getMonth() + 3) / 3);
					let from_date;
					
					// Set date range based on current quarter
					switch(quarter) {
						case 1: // Jan-Mar
							from_date = frappe.datetime.year_start(today);
							break;
						case 2: // Apr-Jun
							from_date = frappe.datetime.add_months(frappe.datetime.year_start(today), 3);
							break;
						case 3: // Jul-Sep
							from_date = frappe.datetime.add_months(frappe.datetime.year_start(today), 6);
							break;
						case 4: // Oct-Dec
							from_date = frappe.datetime.add_months(frappe.datetime.year_start(today), 9);
							break;
					}
					
					let to_date = frappe.datetime.add_months(from_date, 3);
					to_date = frappe.datetime.add_days(to_date, -1);
					
					frappe.query_report.set_filter_value('from_date', from_date);
					frappe.query_report.set_filter_value('to_date', to_date);
				} 
				else if (periodicity === "Yearly") {
					let fiscal_year = frappe.defaults.get_user_default("fiscal_year");
					
					if (!fiscal_year) {
						// If no fiscal year is set, use calendar year
						let from_date = frappe.datetime.year_start(today);
						let to_date = frappe.datetime.year_end(today);
						
						frappe.query_report.set_filter_value('from_date', from_date);
						frappe.query_report.set_filter_value('to_date', to_date);
					} else {
						frappe.db.get_value("Fiscal Year", fiscal_year, ["year_start_date", "year_end_date"], (r) => {
							if (r) {
								frappe.query_report.set_filter_value('from_date', r.year_start_date);
								frappe.query_report.set_filter_value('to_date', r.year_end_date);
							}
						});
					}
				}
				else {
					// Monthly - default to current month
					let from_date = frappe.datetime.month_start(today);
					let to_date = frappe.datetime.month_end(today);
					
					frappe.query_report.set_filter_value('from_date', from_date);
					frappe.query_report.set_filter_value('to_date', to_date);
				}
			}
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_start(frappe.datetime.get_today()),
			reqd: 1,
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.month_end(frappe.datetime.get_today()),
			reqd: 1,
		},
		{
			fieldname: "asset",
			label: __("Asset"),
			fieldtype: "Link",
			options: "Asset",
		},
		{
			fieldname: "asset_category",
			label: __("Asset Category"),
			fieldtype: "Link",
			options: "Asset Category",
		},
		{
			fieldname: "finance_book",
			label: __("Finance Book"),
			fieldtype: "Link",
			options: "Finance Book",
		},
		// {
		// 	fieldname: "include_default_book_assets",
		// 	label: __("Include Default FB Assets"),
		// 	fieldtype: "Check",
		// 	default: 1,
		// },
	]
  };